controller.go

```diff
@@ -432,419 +407,261 @@ func (c *Controller) syncHandler(key string) error {
-				OS:       "projects/debian-cloud/global/images/debian-9-stretch-v20191121",
-				Plan:     "f1-micro",
-				UserData: userData,
-				Additional: map[string]string{
-					"projectid":     c.infraConfig.ProjectID,
-					"zone":          c.infraConfig.Zone,
-					"firewall-name": firewallRuleName,
-					"firewall-port": strconv.Itoa(inletsPort),
-				},
-			})
-
-			if err != nil {
-				return err
-			}
-			id = res.ID
-
-		} else if c.infraConfig.Provider == "ec2" {
-			provisioner, _ := provision.NewEC2Provisioner(c.infraConfig.Region, c.infraConfig.GetAccessKey(), c.infraConfig.GetSecretKey())
-
-			inletsPort := inletsControlPort
-
-			if c.infraConfig.UsePro() {
-				inletsPort = inletsProControlPort
-			}
-
-			userData := makeUserdata(tunnel.Spec.AuthToken, c.infraConfig.UsePro(), tunnel.Spec.ServiceName)
-
-			res, err := provisioner.Provision(provision.BasicHost{
-				Name:     tunnel.Name,
-				OS:       "ubuntu/images/hvm-ssd/ubuntu-xenial-16.04-amd64-server-20191114",
-				Plan:     "t3.micro",
-				UserData: base64.StdEncoding.EncodeToString([]byte(userData)),
-				Additional: map[string]string{
-					"inlets-port": strconv.Itoa(inletsPort),
-				},
-			})
-
-			if err != nil {
-				return err
-			}
-			id = res.ID
-		} else if c.infraConfig.Provider == "civo" {
-			provisioner, _ := provision.NewCivoProvisioner(c.infraConfig.GetAccessKey())
-
-			userData := makeUserdata(tunnel.Spec.AuthToken, c.infraConfig.UsePro(), tunnel.Spec.ServiceName)
-
-			res, err := provisioner.Provision(provision.BasicHost{
-				Name:       tunnel.Name,
-				OS:         "811a8dfb-8202-49ad-b1ef-1e6320b20497",
-				Plan:       "g2.small",
-				Region:     c.infraConfig.Region,
-				UserData:   userData,
-				Additional: map[string]string{},
-			})
-
-			if err != nil {
-				return err
-			}
-			id = res.ID
-		} else if c.infraConfig.Provider == "linode" {
-			provisioner, _ := provision.NewLinodeProvisioner(c.infraConfig.GetAccessKey())
-
-			userData := makeUserdata(tunnel.Spec.AuthToken, c.infraConfig.UsePro(), tunnel.Spec.ServiceName)
-
-			res, err := provisioner.Provision(provision.BasicHost{
-				Name:       tunnel.Name,
-				OS:         "linode/ubuntu16.04lts", // https://api.linode.com/v4/images
-				Plan:       "g6-nanode-1",           // https://api.linode.com/v4/linode/types
-				Region:     c.infraConfig.Region,
-				UserData:   userData,
-				Additional: map[string]string{},
-			})
-
-			if err != nil {
-				return err
-			}
-			id = res.ID
-
-		} else {
-			return fmt.Errorf("unsupported provider: %s", c.infraConfig.Provider)
+		host := getHostConfig(c, tunnel)
+		res, err := provisioner.Provision(host)
+		if err != nil {
+			return err
 		}
-
 		log.Printf("Provisioning call took: %fs\n", time.Since(start).Seconds())
+		// End of Provisioning Host
 
+		// Updating Status
+		err = c.updateTunnelProvisioningStatus(tunnel, "provisioning", res.ID, "")
 		if err != nil {
-			err = c.updateTunnelProvisioningStatus(tunnel, "error", "", "")
```

Comment:
We used to set an error status, why did that change?

Reply:
That part of code was actually a dead code in original code. So the error status setting never happened.

![image](https://user-images.githubusercontent.com/138433/87957411-cda84c80-cae2-11ea-941a-52a3b3e018bf.png)

Now this `error != nil` on line 428 is actually to handle the error of previous "updatingTunnel provisioning" failed, meaning if we call it to set the error, it will probably fail as well. So simply returning the error is more making sense in my opinion.

This logic is working fine. Per my testing, simply returning the error is enough that it will retry. (I tried to turn off my network to testing the retry scenario)